File: leetcode_python/02.05.py
# 面试题 02.05. 链表求和
# https://leetcode-cn.com/problems/sum-lists-lcci/
from typing import List
from collections import defaultdict
from collections import deque
from math import inf
import bisect
import heapq
'''
@File : 02.05.py
@Time : 2022/01/04 19:43:19
@Author : YuMin Zhang
@State : Indepeedent
@Thinking :
@Tag : Medium
'''
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        carry = 0
        ans = ListNode(0)
        tmp_node = ans
        while l1 and l2:
            num = l2.val + l1.val + carry 
            carry = num // 10
            tmp = ListNode(num % 10)
            tmp_node.next = tmp
            tmp_node = tmp_node.next
            l1 = l1.next
            l2 = l2.next        
        if l1:
            while l1 and carry != 0:
                num = l1.val + carry
                carry = num // 10
                l1.val = num % 10
                tmp_node.next = l1
                tmp_node = tmp_node.next
                l1 = l1.next
            if l1:
                tmp_node.next = l1
        elif l2:
            while l2 and carry != 0:
                num = l2.val + carry
                carry = num // 10
                l2.val = num % 10
                tmp_node.next = l2
                tmp_node = tmp_node.next
                l2 = l2.next
            if l2:
                tmp_node.next = l2

        if carry != 0:
            tmp_node.next = ListNode(carry)
        return ans.next

File: leetcode_c_plus_plus/02.05.cpp
//
//
#include<map>
#include<set>
#include<vector>
#include<cmath>
#include<numeric>
#include<unordered_map>
using namespace std;
#define INT_MAX 2147483647
#define INT_MIN (-INT_MAX - 1)
/**
@File : 02.05.cpp
@Time : 2022/01/19 22:35:27
@Author : YuMin Zhang
@State : Indepeedent | Half-Depedent | Depedent 
@Thinking :
@Tag : Easy | Medium | Hard
**/
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {

        ListNode* dummy=new ListNode(0);//哑节点
        ListNode* cur=dummy;
        int add=0;
        while(l1||l2||add){
            add+=(l1?l1->val:0)+(l2?l2->val:0);//进位的值+l1节点值+l2节点值
            cur->next=new ListNode(add%10);//当前节点的值 add%10
            add/=10;//获得进位的值 add/10
            cur=cur->next;
            l1=l1?l1->next:nullptr;//l1移动下一个节点(如果不为空)
            l2=l2?l2->next:nullptr;//l2移动下一个节点(如果不为空)
        }
        return dummy->next;
    }
};